### ISSUE TITLE
Sentry.NLog and ${aspnet-user-identity} as user of event

### ISSUE THREAD

**Issue by oddeirik** [getsentry/sentry-dotnet#322]
Please mark the type framework used:

- [X] ASP.NET MVC
- [ ] ASP.NET Web API (OWIN)
- [ ] ASP.NET Core
- [ ] WPF
- [ ] WinForms
- [ ] Xamarin
- [ ] Other:

Please mark the type of the runtime used:

- [X] .NET Framework
- [ ] Mono
- [ ] .NET Core

* Version:

Please mark the NuGet packages used:

- [X] Sentry
- [ ] Sentry.Serilog
- [X] Sentry.NLog
- [ ] Sentry.Log4Net
- [ ] Sentry.Extensions.Logging
- [ ] Sentry.AspNetCore

* Version: 1.2.0

The TLDR version is basically; I want to set the username in the Sentry NLog events to ${aspnet-user-identity}, any idea how this can be done?

The longer version:
SDK is initialized in Global.asax.cs, and on unhandled exceptions we're setting a scope to include the currently authenticated user (logged in via Azure AD). This works as expected for unhandled exceptions. (I've tried using `sendDefaultPii` here, but that only included the service user for the application, which isn't of interest to us.)

We also have a Sentry NLog target configured for log events containing exceptions where we don't necessarily want to throw an exception, but we want to send it to Sentry. Some of these exceptions are happening within the scope/request of an authenticated user, but for some reason this information is not included in the Sentry event.

I've tried setting `sendDefaultPii` in the NLog configuration, with no luck. I've also tried sending a custom tag with the user/username explicitly set:

`<tag name="user" layout="username:${aspnet-user-identity}" />`

However, this information is not included in the event being sent, so it's either ignored or filtered out. I've tried adding ${aspnet-user-identity} to log layout and the currently logged in user is being shown as expected there. Is there a way to set the scope of a log event with a specific username, similar to how we're doing it in Global.asax.cs?

**Comment by josh-degraw** [comment]
So you're saying that the `aspnet-user-identity` variable works when you use it as the log message layout, but it gets dropped if it's included as a tag? That certainly is strange. I assume you're already referencing `NLog.Web` and including it in the NLog config file?

If you could post a minimal sample of your nlog config and some of the code of setting up the scope that could help nail down the problem.

Have you tried setting the username etc in code e.g.:
```c#
 SentrySdk.ConfigureScope(s => s.User.Username = theUserNameEtc);
```
?
If none of that proves helpful, there are a few other ways you can set the variable for NLog to use. You could use one of NLog's [diagnostic context](https://github.com/NLog/NLog/wiki/Context) classes (e.g. `ndlc`) or [variables](https://github.com/NLog/NLog/wiki/Var-Layout-Renderer), although both are much less ideal than having it just work, they could work as a temporary solution.

**Comment by oddeirik** [comment]
Yes, that's correct, it works in the log message, but not when set in a tag called user:
`<tag name="user" layout="username:${aspnet-user-identity}" />`

I've looked at the JSON from events and there's no mention of a user tag anywhere in those, so it looks like it's either being filtered or cleared out, either on send or when received in Sentry.

The NLog Sentry configuration used:

```
<target xsi:type="Sentry" name="sentry"
	dsn="<REDACTED>"
	layout="${longdate} ${uppercase:${level}} [${logger}] [${aspnet-user-identity}] ${aspnet-item:variable=requestid} - ${message} ${exception:format=tostring}"
	breadcrumbLayout="${longdate} ${uppercase:${level}} [${logger}] ${aspnet-item:variable=requestid} - ${message} ${exception:format=tostring}"
	minimumBreadcrumbLevel="Debug"
	minimumEventLevel="Error">

	<options
		ignoreEventsWithNoException="true"
		environment="dev"
		attachStacktrace="true"
		sendDefaultPii="true"
		isEnvironmentUser="false"
		shutdownTimeoutSeconds="5"
		initializeSdk="false"
	>
		<includeEventDataOnBreadcrumbs>true</includeEventDataOnBreadcrumbs>
	</options>

	<tag name="logger" layout="${logger}" />
	<tag name="user" layout="username:${aspnet-user-identity}" />
</target>
```

For unhandled exceptions we set the scope with the authenticated user (if there is one). This is done in the Application_Error method in Global.asax.cs. This works as expected, and user information is included in those cases.

But for log events user information is missing, for example in an ASP.NET base controller class there is exception handling which only logs the exceptions and I don't want to rethrow. These exceptions happen within a user request scope, so it seems weird that the user information is not included then. We could of course set the Sentry scope for those cases, but there are other places where we're just logging the exceptions and I was hoping to avoid setting the scope explicitly in every place we're doing Logger.Error(exception) or similar 😃

**Comment by josh-degraw** [comment]
My guess here would be that it has something to do with the `user` key being reserved or something to that effect. 
So ideally you're wanting to tell sentry the actual identity of the user via the NLog config, rather than just a simple event tag, correct? If that's the case it might make sense to add another config option to the NLog integration as for how to assign user info (and probably other contexts, too) in this manner.

**Comment by oddeirik** [comment]
> So ideally you're wanting to tell sentry the actual identity of the user via the NLog config, rather than just a simple event tag, correct?

Yes, that's pretty much it 😃 

> If that's the case it might make sense to add another config option to the NLog integration as for how to assign user info (and probably other contexts, too) in this manner.

Having an option to configure more options via Nlog configuration, almost like a scope, could be really useful.

**Comment by josh-degraw** [comment]
@bruno-garcia I'm gonna try and tackle this if that's okay. In the meantime, would you mind creating an `NLog` label for issues? I'm certainly down to contribute to more than just `Sentry.NLog` but it would be nice to be able to filter by that.

**Comment by bruno-garcia** [comment]
@josh-degraw great! Sure, will do.

**Comment by bruno-garcia** [comment]
@josh-degraw you actually have permission to do that. Also to assign yourself to the issue and more.

**Comment by josh-degraw** [comment]
@bruno-garcia I tried to but it doesn't seem like I actually can for some reason.

**Comment by snakefoot** [comment]
Resolved with #336 ?

**Comment by oddeirik** [comment]
Yes, we've upgraded Sentry packages to v2.0.1 and with some additional log target configuration:
`<user username="${aspnet-user-identity}" ipAddress="${aspnet-request-ip}" />`,
the username (and IP address in this case) is now included as expected. Thanks! 😃
